### PR TITLE
[X86][MC] Support Enc/Dec for EGPR for promoted CRC32

### PR DESCRIPTION
**Pull request by XinWang10**
R16-R31 was added into GPRs in https://github.com/llvm/llvm-project/pull/70958,
This patch supports the encoding/decoding for promoted CRC32 instruction in EVEX space.

RFC: https://discourse.llvm.org/t/rfc-design-for-apx-feature-egpr-and-ndd-support/73031/4

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrSSE.td [hunk 1]
```diff
@@ -6698,6 +6698,30 @@ let Constraints = "$src1 = $dst" in {
...
   }
+
+  def CRC32r32m8_EVEX  : SS42I_crc32m<0xF0, "crc32{b}", GR32, i8mem,
+                                      int_x86_sse42_crc32_32_8>, EVEX, T_MAP4, PS;
+  def CRC32r32r8_EVEX  : SS42I_crc32r<0xF0, "crc32{b}", GR32, GR8,
+                                      int_x86_sse42_crc32_32_8>, EVEX, T_MAP4, PS;
+  def CRC32r32m16_EVEX : SS42I_crc32m<0xF1, "crc32{w}", GR32, i16mem,
+                                      int_x86_sse42_crc32_32_16>, EVEX, T_MAP4, PD;
+  def CRC32r32r16_EVEX : SS42I_crc32r<0xF1, "crc32{w}", GR32, GR16,
+                                      int_x86_sse42_crc32_32_16>, EVEX, T_MAP4, PD;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrSSE.td:6709]: One question: When do we need set `NoCD8`?

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSSE.td:6709]: EVEX encoding would set CD8_Scale by default, which could interfere compiler when encoding disp8, like 
crc32w 123(%r28,%r29,4), %r18d, we expect compiler could encode the 123 in one byte, but if we have non-zero CD8_Scare, we still need 4 bytes.

File: llvm/lib/Target/X86/X86InstrSSE.td [hunk 2]
```diff
@@ -6698,6 +6698,30 @@ let Constraints = "$src1 = $dst" in {
     def CRC32r64r8 : SS42I_crc32r<0xF0, "crc32{b}", GR64, GR8,
                                    null_frag>, REX_W;
   }
+
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrSSE.td:6701]: @KanRobert said we need `In64BitMode` even for assembler?

File: llvm/test/MC/X86/apx/crc32-att.s [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# RUN: llvm-mc -triple x86_64 --show-encoding %s | FileCheck %s
+
+# CHECK: crc32b	%r16b, %r22d
```
Comment (KanRobert) [inline llvm/test/MC/X86/apx/crc32-att.s:3]: Missing test for code like `{evex} crc32b %eax, %ebx`

File: llvm/lib/Target/X86/X86InstrSSE.td (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrSSE.td:6701]: Yes.

File: llvm/lib/Target/X86/X86InstrSSE.td (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrSSE.td:6709]: So why you don't add NoCD8 here?

File: llvm/lib/Target/X86/X86InstrSSE.td (hunk 2 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSSE.td:6701]: Will add.

File: llvm/lib/Target/X86/X86InstrSSE.td (hunk 1 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSSE.td:6709]: Missing, would add it back.

File: llvm/test/MC/X86/apx/crc32-att.s (hunk 3 above)
Comment (XinWang10) [reply llvm/test/MC/X86/apx/crc32-att.s:3]: OK, then we also miss test for other promoted instrs, I just test egpr for them.

File: llvm/lib/Target/X86/X86InstrSSE.td [hunk 4]
```diff
@@ -6699,28 +6699,30 @@ let Constraints = "$src1 = $dst" in {
...
+                                      int_x86_sse42_crc32_32_8>, EVEX, NoCD8, T_MAP4, PS;
   def CRC32r32m16_EVEX : SS42I_crc32m<0xF1, "crc32{w}", GR32, i16mem,
-                                      int_x86_sse42_crc32_32_16>, EVEX, T_MAP4, PD;
+                                      int_x86_sse42_crc32_32_16>, EVEX, NoCD8, T_MAP4, PD;
   def CRC32r32r16_EVEX : SS42I_crc32r<0xF1, "crc32{w}", GR32, GR16,
-                                      int_x86_sse42_crc32_32_16>, EVEX, T_MAP4, PD;
+                                      int_x86_sse42_crc32_32_16>, EVEX, NoCD8, T_MAP4, PD;
   def CRC32r32m32_EVEX : SS42I_crc32m<0xF1, "crc32{l}", GR32, i32mem,
-                                      int_x86_sse42_crc32_32_32>, EVEX, T_MAP4, PS;
+                                      int_x86_sse42_crc32_32_32>, EVEX, NoCD8, T_MAP4, PS;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrSSE.td:6712]: Move `NoCD8` to `SS42I_crc32r/m` ?

File: llvm/lib/Target/X86/X86InstrSSE.td [hunk 5]
```diff
@@ -6699,28 +6699,30 @@ let Constraints = "$src1 = $dst" in {
                                    null_frag>, REX_W;
   }
 
+  let Predicates = [In64BitMode] in {
   def CRC32r32m8_EVEX  : SS42I_crc32m<0xF0, "crc32{b}", GR32, i8mem,
-                                      int_x86_sse42_crc32_32_8>, EVEX, T_MAP4, PS;
+                                      int_x86_sse42_crc32_32_8>, EVEX, NoCD8, T_MAP4, PS;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrSSE.td:6704]: Remove PS, it's redundant.

File: llvm/lib/Target/X86/X86InstrSSE.td [hunk 6]
```diff
@@ -6699,28 +6699,30 @@ let Constraints = "$src1 = $dst" in {
                                    null_frag>, REX_W;
   }
 
+  let Predicates = [In64BitMode] in {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrSSE.td:6702]: This `let` is incorrect. You will override the default predicate `[HasCRC32]`

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSSE.td:6702]: Nice catch, will fix.

File: llvm/lib/Target/X86/X86InstrSSE.td (hunk 5 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrSSE.td]: It's needed, base class CRC32I has defined XD.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**